### ISSUE TITLE
update documentation for VariableDeclarationUsageDistance with allowedDistance = 0

### ISSUE THREAD

**Issue by vanniktech** [checkstyle/checkstyle#3665]
Given this check is activated:

```xml
<module name="VariableDeclarationUsageDistance">
  <property name="allowedDistance" value="0"/>
</module>
```

it'll fail on this code:

```java
long result = dateTime != null ? dateTime.hashCode() : 0L;
result = 31L * result + (shift != null ? shift.hashCode() : 0L);
return result;
```

with this error message:

> Distance between variable 'result' declaration and its first usage is 1, but allowed 0.  Consider to make that variable as final if you still need to store its value in advance (before method calls that might do side effect on original value). [VariableDeclarationUsageDistance]

Is this really wanted? I'd expect it not to fail since the next usage is really one line below it.

**Comment by rnveach** [comment]
@vannitech I believe 0 is an invalid value. It basically says all variables can only reference itself. See ecample below.
Try using a value of 1 instead of 0.

````
$ cat TestClass.java
public class TestClass {
    void method() {
long result = dateTime != null ? dateTime.hashCode() : 0L;
return result;
    }
}

$ cat TestConfig.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
          "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">

<module name="Checker">
    <property name="charset" value="UTF-8"/>

    <module name="TreeWalker">
<module name="VariableDeclarationUsageDistance">
  <property name="allowedDistance" value="0"/>
</module>
    </module>
</module>

$ java -jar checkstyle-7.3-all.jar -c TestConfig.xml TestClass.java
Starting audit...
[ERROR] TestClass.java:3: Distance between variable 'result' declaration and its first usage is 1, but allowed 0.  Consider to make that variable as final if you still need to store its value in advance (before method calls that might do side effect on original value). [VariableDeclarationUsageDistance]
Audit done.
Checkstyle ends with 1 errors.
````

**Comment by romani** [comment]
value `0` does not make any sense, unfortunately it is not described in our documentation - http://checkstyle.sourceforge.net/config_coding.html#VariableDeclarationUsageDistance

in scope of this bug, documentation need to be updated.

**Comment by rnveach** [comment]
Since the property is a numeric field, it allows negative numbers too.
When we update documentation, it should be to discourage using anything less than 1.

**Comment by vanniktech** [comment]
Updating the documentation sounds good to me. Might be even worth forbidding negative values.

**Comment by romani** [comment]
doc is extended.
issue is closed
